### PR TITLE
Backend/feat/clever-tap-backend

### PR DESCRIPTION
**Pull request by prashant601**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description

Rebased this branch with KV,
here's the previous PR:
https://github.com/nammayatri/nammayatri/pull/1769

### Additional Changes

- [x] This PR modifies the database schema (database migration added)
- [x] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [x] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable
- [ ] No leak detected in leakcanary

### CODE REVIEW

File: Backend/app/kafka-consumers/src/Consumer/CustomerStats/Processor.hs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+              _ -> pure ()
+          BookingCancelled -> do
+            unless (isNotBackfilled personStats) do
+              let bookingId = cast payload.bId
+              -- bookingCancellationReason <- Esq.runInReplica $ QBCR.findByRideBookingId bookingId >>= fromMaybeM (BookingNotFound bookingId.getId)
+              bookingCancellationReason <- runInReplica $ QBCR.findByRideBookingId bookingId >>= fromMaybeM (BookingNotFound bookingId.getId)
+              -- Esq.runNoTransaction $ do
+              if bookingCancellationReason.source == SBCR.ByUser
+                then QP.incrementUserCancelledRidesCount personId
+                else QP.incrementDriverCancelledRidesCount personId
```
Comment (0utkarsh) [inline Backend/app/kafka-consumers/src/Consumer/CustomerStats/Processor.hs:75]: same applies here

File: Backend/app/kafka-consumers/src/Consumer/CustomerStats/Processor.hs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+                  when (SP.isWeekend createdAt minuteDiffFromUTC) do QP.incrementWeekendRidesCount personId
+                  unless (SP.isWeekend createdAt minuteDiffFromUTC) do QP.incrementWeekdayRidesCount personId
+                  when (SP.checkMorningPeakInWeekday createdAt minuteDiffFromUTC) do QP.incrementMorningPeakRidesCount personId
+                  when (SP.checkEveningPeakInWeekday createdAt minuteDiffFromUTC) do QP.incrementEveningPeakRidesCount personId
+                  when (SP.checkWeekendPeak createdAt minuteDiffFromUTC) do QP.incrementWeekendPeakRidesCount personId
+                  when (not (SP.checkEveningPeakInWeekday createdAt minuteDiffFromUTC) && not (SP.checkMorningPeakInWeekday createdAt minuteDiffFromUTC) && not (SP.checkWeekendPeak createdAt minuteDiffFromUTC)) do QP.incrementOffpeakRidesCount personId
+                  QP.incrementCompletedRidesCount personId
+              _ -> pure ()
+          BookingCancelled -> do
+            unless (isNotBackfilled personStats) do
```
Comment (0utkarsh) [inline Backend/app/kafka-consumers/src/Consumer/CustomerStats/Processor.hs:68]: This condition is in both cases, so can be present at the top

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/PersonStats.hs [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+module Domain.Action.UI.PersonStats where
+
+import Data.Time hiding (getCurrentTime)
+import qualified Domain.Types.Merchant as Merchant
+import qualified Domain.Types.Person as DP
+-- import Kernel.Storage.Esqueleto (runInReplica)
+import Kernel.Beam.Functions
+import Kernel.External.Encryption
+import Kernel.Prelude
+-- import qualified Kernel.Storage.Esqueleto as Esq
```
Comment (0utkarsh) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/PersonStats.hs:24]: Can you remove these comments(are present at other places too)

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/PersonStats.hs [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+      weekendRidesRate =
+        fromIntegral personStats.weekendRides
+          / fromIntegral personStats.completedRides
+      weekdayRidesRate =
+        fromIntegral personStats.weekdayRides
+          / fromIntegral personStats.completedRides
+      weekendPeakRideRate =
+        fromIntegral personStats.weekendPeakRides
+          / fromIntegral personStats.completedRides
+      daysFromRegDate :: Int = roundToIntegral $ diffUTCTime now person.createdAt / 60 / 60 / 24
```
Comment (0utkarsh) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/PersonStats.hs:119]: Any specific reason to do it this way - `/ 60 / 60 / 24` ?

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Registration.hs [hunk 5]
```diff
@@ -436,12 +438,14 @@ getRegistrationTokenE :: EsqDBFlow m r => Id SR.RegistrationToken -> m SR.Regist
 getRegistrationTokenE tokenId =
   RegistrationToken.findById tokenId >>= fromMaybeM (TokenNotFound $ getId tokenId)
 
-createPerson :: (EncFlow m r, EsqDBFlow m r, DB.EsqDBReplicaFlow m r, Redis.HedisFlow m r, CacheFlow m r) => AuthReq -> Text -> Maybe Text -> Maybe Version -> Maybe Version -> Id DMerchant.Merchant -> m SP.Person
+createPerson :: (EncFlow m r, EsqDBFlow m r, DB.EsqDBReplicaFlow m r, Redis.HedisFlow m r, CacheFlow m r, CoreMetrics m) => AuthReq -> Text -> Maybe Text -> Maybe Version -> Maybe Version -> Id DMerchant.Merchant -> m SP.Person
```
Comment (0utkarsh) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Registration.hs:441]: Why this constraint?

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Registration.hs [hunk 6]
```diff
@@ -436,12 +438,14 @@ getRegistrationTokenE :: EsqDBFlow m r => Id SR.RegistrationToken -> m SR.Regist
...
 createPerson req mobileNumber notificationToken mbBundleVersion mbClientVersion merchantId = do
   person <- buildPerson req mobileNumber notificationToken mbBundleVersion mbClientVersion merchantId
+  createPersonStats <- makePersonStats person
   -- DB.runTransaction $ do
   _ <- Person.create person
   _ <- QDFS.create $ makeIdlePersonFlowStatus person
+  _ <- QPS.create createPersonStats
```
Comment (0utkarsh) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Registration.hs:448]: Is it necessary to create it here? You have to always check `(all = 0)` before filling data when you could just check if it exists or not. When stats API is called for a new person, we could show all 0. This could be done to save some data for users who just register and don't use but fine, you may choose to ignore it also depending on what you like more.

File: Backend/app/rider-platform/rider-app/Main/src/Storage/Beam/Merchant.hs [hunk 7]
```diff
@@ -199,6 +201,7 @@ merchantTMod =
       signatureExpiry = B.fieldNamed "signature_expiry",
       updatedAt = B.fieldNamed "updated_at",
       createdAt = B.fieldNamed "created_at",
+      timeDiffFromUtc = B.fieldNamed "time_diff_from_utc",
```
Comment (0utkarsh) [inline Backend/app/rider-platform/rider-app/Main/src/Storage/Beam/Merchant.hs:204]: Why is this added here when you could get it from `transporterConfig` table?

**Review by 0utkarsh — CHANGES_REQUESTED** [review]

File: Backend/app/rider-platform/rider-app/Main/src/SharedLogic/Person.hs [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+  cancelledBookingIds <- runInReplica $ QBooking.findAllCancelledBookingIdsByRider personId
+  incrementUserCancelledRidesBy <- runInReplica $ QBCR.countCancelledBookingsByRiderId cancelledBookingIds DBCR.ByUser
+  let incrementDriverCancelledRidesBy = length cancelledBookingIds - incrementUserCancelledRidesBy
+  completedRides <- runInReplica QRide.findAllCompletedRides
+  merchant <- CQMerchant.findById merchantId >>= fromMaybeM (MerchantDoesNotExist merchantId.getId)
+  let minuteDiffFromUTC = (merchant.timeDiffFromUtc.getSeconds) `div` 60
+  now <- getCurrentTime
+  let completedRidesCnt = length completedRides
+  let (incrementWeekendRidesBy, incrementWeekdayRidesBy) = countWeekdaysAndWeekendsRide completedRides minuteDiffFromUTC
+      incrementEveningPeakRides = getWeekdayEveningPeakRides completedRides minuteDiffFromUTC
```
Comment (0utkarsh) [inline Backend/app/rider-platform/rider-app/Main/src/SharedLogic/Person.hs:49]: You can have a table which can keep track of these peak timings per merchant. It will give you the flexibility of

1. Adding a new peak time without code change
2. Updating peak time for a specific type
3. Having different peak times for different merchants (e.g. France morning's peak could be different than Kolkata)

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/PersonStats.hs (hunk 3 above)
Comment (prashant601) [reply Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/PersonStats.hs:24]: I haven't removed these comments  because KV. branch also have comments of Esq and they want them for reference, I've  discussed with them before leaving the comments

File: Backend/app/rider-platform/rider-app/Main/src/Storage/Beam/Merchant.hs (hunk 7 above)
Comment (prashant601) [reply Backend/app/rider-platform/rider-app/Main/src/Storage/Beam/Merchant.hs:204]: hemant told me to add this in merchant table as it was in driver side

File: Backend/app/kafka-consumers/src/Consumer/CustomerStats/Processor.hs (hunk 2 above)
Comment (prashant601) [reply Backend/app/kafka-consumers/src/Consumer/CustomerStats/Processor.hs:68]: done

File: Backend/app/kafka-consumers/src/Consumer/CustomerStats/Processor.hs (hunk 1 above)
Comment (prashant601) [reply Backend/app/kafka-consumers/src/Consumer/CustomerStats/Processor.hs:75]: changed

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/PersonStats.hs (hunk 4 above)
Comment (prashant601) [reply Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/PersonStats.hs:119]: not really, Alexander worked on this task, so some changes are not mine .. this is one of them

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Registration.hs (hunk 5 above)
Comment (prashant601) [reply Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Registration.hs:441]: removed it

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Registration.hs (hunk 6 above)
Comment (prashant601) [reply Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Registration.hs:448]: Sure

File: Backend/app/rider-platform/rider-app/Main/src/SharedLogic/Person.hs [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+import qualified Storage.Queries.Booking as QBooking
+import qualified Storage.Queries.BookingCancellationReason as QBCR
+import qualified Storage.Queries.Ride as QRide
+import Tools.Metrics (CoreMetrics)
+
+backfillPersonStats :: (EsqDBFlow m r, EsqDBReplicaFlow m r, SCC.CacheFlow m r, CoreMetrics m) => Id DP.Person -> Id Merchant.Merchant -> m DPS.PersonStats
+backfillPersonStats personId merchantId = do
+  cancelledBookingIds <- runInReplica $ QBooking.findAllCancelledBookingIdsByRider personId
+  incrementUserCancelledRidesBy <- runInReplica $ QBCR.countCancelledBookingsByRiderId cancelledBookingIds DBCR.ByUser
+  let incrementDriverCancelledRidesBy = length cancelledBookingIds - incrementUserCancelledRidesBy
```
[the remark this answers is not in the document]
Comment (prashant601) [reply Backend/app/rider-platform/rider-app/Main/src/SharedLogic/Person.hs:42]: corrected it

File: Backend/app/rider-platform/rider-app/Main/src/SharedLogic/Person.hs (hunk 8 above)
Comment (prashant601) [reply Backend/app/rider-platform/rider-app/Main/src/SharedLogic/Person.hs:49]: for now , this approach was discussed that's why I did it that way

File: Backend/app/rider-platform/rider-app/Main/src/SharedLogic/Person.hs [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+  completedRides <- runInReplica QRide.findAllCompletedRides
+  merchant <- CQMerchant.findById merchantId >>= fromMaybeM (MerchantDoesNotExist merchantId.getId)
+  let minuteDiffFromUTC = (merchant.timeDiffFromUtc.getSeconds) `div` 60
+  now <- getCurrentTime
+  let completedRidesCnt = length completedRides
+  let (incrementWeekendRidesBy, incrementWeekdayRidesBy) = countWeekdaysAndWeekendsRide completedRides minuteDiffFromUTC
+      incrementEveningPeakRides = getWeekdayEveningPeakRides completedRides minuteDiffFromUTC
+      incrementMorningPeakRides = getWeekdayMorningPeakRides completedRides minuteDiffFromUTC
+      incrementOffPeekRidesBy = completedRidesCnt - (incrementMorningPeakRides + incrementEveningPeakRides)
```
[the remark this answers is not in the document]
Comment (prashant601) [reply Backend/app/rider-platform/rider-app/Main/src/SharedLogic/Person.hs:51]: modified

File: Backend/app/rider-platform/rider-app/Main/src/SharedLogic/Person.hs [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+        DPS.PersonStats
+          { personId = personId,
+            userCancelledRides = incrementUserCancelledRidesBy,
+            driverCancelledRides = incrementDriverCancelledRidesBy,
+            completedRides = completedRidesCnt,
+            weekendRides = incrementWeekendRidesBy,
+            weekdayRides = incrementWeekdayRidesBy,
+            offPeakRides = incrementOffPeekRidesBy,
+            eveningPeakRides = incrementEveningPeakRides,
+            morningPeakRides = incrementMorningPeakRides,
```
[the remark this answers is not in the document]
Comment (prashant601) [reply Backend/app/rider-platform/rider-app/Main/src/SharedLogic/Person.hs:62]: done

File: Backend/app/rider-platform/rider-app/Main/src/Storage/Queries/BookingCancellationReason.hs [hunk 12]
```diff
@@ -70,6 +72,32 @@ upsert cancellationReason = do
...
+--       where_ $
+--         bookingCancellationReason ^. BookingCancellationReasonSource ==. val cancellationSource
+--           &&. bookingCancellationReason ^. BookingCancellationReasonBookingId `in_` valList (toKey <$> bookingIds)
+--       return (countRows :: SqlExpr (Esq.Value Int))
+--   where
+--     mkCount [counter] = counter
+--     mkCount _ = 0
+
+countCancelledBookingsByRiderId :: (L.MonadFlow m, Log m) => [Id Booking] -> CancellationSource -> m Int
+countCancelledBookingsByRiderId bookingIds cancellationSource = do
```
[the remark this answers is not in the document]
Comment (prashant601) [reply Backend/app/rider-platform/rider-app/Main/src/Storage/Queries/BookingCancellationReason.hs:89]: modified

File: Backend/app/rider-platform/rider-app/Main/src/Storage/Beam/Merchant.hs (hunk 7 above)
Comment (hkmangla) [reply Backend/app/rider-platform/rider-app/Main/src/Storage/Beam/Merchant.hs:204]: This is rider @0utkarsh . We don't have transporterConfig in rider app.

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/PersonStats.hs (hunk 3 above)
Comment (0utkarsh) [reply Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/PersonStats.hs:24]: Sure, you can have it in the queries for reference but you can remove from other places and also these imports.

File: Backend/app/rider-platform/rider-app/Main/src/SharedLogic/Person.hs (hunk 8 above)
Comment (0utkarsh) [reply Backend/app/rider-platform/rider-app/Main/src/SharedLogic/Person.hs]: Sure, if you want to make this change later, we can do it in next iteration also but I think this will be required eventually.
